Add Retry Interval to REQUEST_ERROR (#1339)
moq-wg/moq-transport: draft-ietf-moq-transport.md

* Error Code: Identifies an integer error code for request failure.

* Retry Interval: The minimum time (in seconds) before the request SHOULD be
sent again. Present only if the Error Code is odd.

* Error Reason: Provides a text description of the request error. See
 {{reason-phrase}}.

The application SHOULD use a relevant error code in REQUEST_ERROR,
as defined below and assigned in {{iana-request-error}}. Most codepoints have
identical meanings for various request types, but some have request-specific
meanings.

Odd error codes indicate the request is retryable with the same parameters at a
later time. If so, the sender of REQUEST_ERROR includes a Retry Interval in the

> message. If it is sending more than one such message within a second or so, it

Reviewer: Do you mean the same error to multiple endpoints, or multiple errors to the same endpoint, or both?

Reviewer: I'd recommend mentioning "across one or more sessions" to make it clear that the second case is included?

> SHOULD jitter the Retry Interval to avoid a flood of retries. The Retry Interval

Reviewer: Do we need more text explaining what "jitter the retry interval" means?

Author: It seems clear to me. propose something if you want it.

If the sender has no information as to when a request is likely to be
successful, it MAY jitter around a default interval of 30 seconds.

If a Retry Interval exceeds the lifetime of a necessary authentication token
used in the request, so that a retry at that time would faile, the sender SHOULD
use a non-retryable error code indicating a new authentication token is needed.

INTERNAL_ERROR:
: An implementation specific or generic error occurred.

UNAUTHORIZED:
: The subscriber is not authorized to perform the requested action on the given
track.

TIMEOUT:
: The subscription could not be completed before an implementation specific
  timeout. For example, a relay could not establish an upstream subscription
  within the timeout.

NOT_SUPPORTED:
: The endpoint does not support the type of request.

MALFORMED_AUTH_TOKEN:
: Invalid Auth Token serialization during registration (see
  {{authorization-token}}).

EXPIRED_AUTH_TOKEN:
: Authorization token has expired ({{authorization-token}}).

Below are errors for use by the publisher. They can appear in response to
SUBSCRIBE, FETCH, TRACK_STATUS, and SUBSCRIBE_NAMESPACE, unless otherwise noted.

> DOES_NOT_EXIST:

Reviewer: Do we need both flavors of DOES_NOT_EXIST?  eg a simple app that gets a SUBSCRIBE for a track it will never publish never wants to see that again.  A relay might simply not have that track "right now".

: The track or namespace is not available at the publisher.

INVALID_RANGE:
: In response to SUBSCRIBE or FETCH, specified Filter or range of Locations
cannot be satisfied.

MALFORMED_TRACK:
: In response to a FETCH, a relay publisher detected
the track was malformed (see {{malformed-tracks}}).

The following are errors for use by the subscriber. They can appear in response
to PUBLISH or PUBLISH_NAMESPACE, unless otherwise noted.

> UNINTERESTED:

Reviewer: Are there both flavors of uninterested?

: The subscriber is not interested in the track or namespace.

Errors below can only be used in response to one message type.

> PREFIX_OVERLAP:

Reviewer: This is sort of retryable if I unsubscribe to a namespace first?  Maybe this is fine.

* Error Code: Identifies an integer error code for request failure.

* Retry Interval: The minimum time (in seconds) before the request SHOULD be
  sent again, plus one. If the value is 0, the request SHOULD NOT be retried.

* Error Reason: Provides a text description of the request error. See
 {{reason-phrase}}.

The application SHOULD use a relevant error code in REQUEST_ERROR,
as defined below and assigned in {{iana-request-error}}. Most codepoints have
identical meanings for various request types, but some have request-specific
meanings.

If a request is retryable with the same parameters at a later time, the sender of
REQUEST_ERROR includes a non-zero Retry Interval in the message. If it is
sending more than one such message within a second or so across one or more

> sessions, it SHOULD apply randomization to each retry interval so that retries

Reviewer 2: ```suggestion
sending more than one such message within a second or so across multiple
sessions, it SHOULD apply randomization to each retry interval so that retries
```

are spread out over time, minimizing the risk of synchronized retry storms.  A

> Retry Interval value of 1 indicates the request can be retried immediately.

Reviewer 3: I think it needs to do randomization even if it is not doing multiple requests. A common problem is Start of Service attacks. With IP phones we regularly see cities of over 10 million people loose power, then all restart at the same time often all with a set top box from of some sort from the same vendor. This forms a Start Of Service attack where the normal mitigations is for the server to be able to respond with retry after and it is more robust to spreading out the load if that can be spread. 

My suggestion would be just make this very specific that the min time to respond was a number between the values returned and 125% of value returned with uniform random distribution in that time. Sure some clients won't bother but the ones that cause SoS attacks will.

INTERNAL_ERROR:
: An implementation specific or generic error occurred. This might be retryable

> or not, depending on the implementation conditions that caused the error.

Reviewer: ```suggestion
: An implementation specific or generic error occurred.
```

* Error Code: Identifies an integer error code for request failure.

* Retry Interval: The minimum time (in seconds) before the request SHOULD be
  sent again, plus one. If the value is 0, the request SHOULD NOT be retried.

* Error Reason: Provides a text description of the request error. See
 {{reason-phrase}}.

The application SHOULD use a relevant error code in REQUEST_ERROR,
as defined below and assigned in {{iana-request-error}}. Most codepoints have
identical meanings for various request types, but some have request-specific
meanings.

If request is retryable with the same parameters at a later time, the sender of
REQUEST_ERROR includes a non-zero Retry Interval in the message. If it is
sending more than one such message within a second or so across one or more
sessions, it SHOULD apply randomization to each retry interval so that retries
are spread out over time, minimizing the risk of synchronized retry storms.  The

> Retry Interval value MAY be one if the request can be retried immediately.

Reviewer: ```suggestion
are spread out over time, minimizing the risk of synchronized retry storms.  A
Retry Interval value of 1 indicates the request can be retried immediately.
```

If the sender has no information as to when a request is likely to be

> successful, it MAY apply randomization around a default interval of 30 seconds.

Reviewer 2: This is ok, but do we need this?

Reviewer: Or make it a SHOULD?

If a Retry Interval exceeds the lifetime of a necessary authentication token
used in the request, so that a retry at that time would fail, the sender SHOULD

> use an error code indicating a new authentication token is needed.

Reviewer: I am inclined to remove this entirely.

> 

Reviewer: ```suggestion
```

| INVALID_JOINING_REQUEST_ID | 0x32 | {{message-request-error}} |
| UNKNOWN_STATUS_IN_RANGE    | 0x33 | {{message-request-error}} |

The range of error codes including 0x100 to 0xffff is reserved for
implementation-speceific codes.

> The range of error codes starting with 0x10000 is reserved for

Reviewer 3: I don't see any reason we need to invent new IANA processes beyond what is in rfc8126
